File: integration/js/openws-ui/ui/utils/endpoint.ts
import type { Endpoint } from '@polytric/openws-spec/types'

export function toUrl(ep: Endpoint) {
    const h = ep.host.trim()
    const pt = ep.port.toString().trim()
    return `${ep.scheme}://${h}${pt ? `:${pt}` : ''}${ep.path.startsWith('/') ? ep.path : `/${ep.path}`}`
}

File: integration/csharp/io.polytric.openws.spec/Runtime/Attributes/Endpoint.cs
using System;

namespace Polytric.OpenWs.Spec.Attributes
{
    /// <summary>
    /// Declares an endpoint associated with a peer role.
    /// Multiple endpoints per role are allowed.
    /// </summary>
    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
    public sealed class EndpointAttribute : Attribute
    {
        public string Scheme { get; }
        public string Host { get; }
        public int Port { get; }
        public string Path { get; }

        public EndpointAttribute(string scheme, string host, int port, string path)
        {
            Scheme = scheme;
            Host = host;
            Port = port;
            Path = path;
        }
    }
}
